# 3 Control Flow
Here we will learn to add more functionality to our code by using conditonal staements and loops. In this lesson we will be going over the many different ways to control the flow of your code.

## 3.1 Conditional Statements
Lets start with a real world example, say a customer is use a mobile app payment app where they spend the credit they have in there wallet. The user wants to make sure that they always have funds so they want the app to automatically transfer funds from their bank account to there mobile wallet when they have less than $5 in credit. This can easily be done in code using an <code>if</code> statenent. An <code>if</code> statement is a conditional statement that runs or skips code based on whether a condition is true or false. The example is below. 


In [1]:
if phone_balance < 5:
    phone_balance += 10
    bank_balance -= 10

NameError: name 'phone_balance' is not defined

Let's break this down.

An <code>if</code> statement starts with the <code>if</code> keyword, followed by the condition to be checked, in this case phone_balance < 5, and then a colon. The condition is specified in a boolean expression that evaluates to either True or False.

After this line is an indented block of code to be executed if that condition is true. Here, the lines that increment phone_balance and decrement bank_balance only execute if it is true that phone_balance is less than 5. If not, the code in this if block is simply skipped.

Use Comparison Operators in Conditional Statements! We have just seen how to run a block of code when a condition is true but say we wanted a different condition to look for when the first condition is false. 


If, Elif, Else
In addition to the if clause, there are two other optional clauses often used with an if statement. For example:


In [ ]:
if season == 'spring':
    print('plant the garden!')
elif season == 'summer':
    print('water the garden!')
elif season == 'fall':
    print('harvest the garden!')
elif season == 'winter':
    print('stay indoors!')
else:
    print('unrecognized season')

The code allows use to check for which season it is and from there we can decide which line of code to run!

1. <code>if</code>: An <code>if</code> statement must always start with an <code>if</code> clause, which contains the first condition that is checked. If this evaluates to True, Python runs the code indented in this if block and then skips to the rest of the code after the <code>if</code> statement.

2. <code>elif</code>: elif is short for "else if." An elif clause is used to check for an additional condition if the conditions in the previous clauses in the if statement evaluate to False. As you can see in the example, you can have multiple elif blocks to handle different situations.

3. <code>else</code>: Last is the else clause, which must come at the end of an <code>if</code> statement if used. This clause doesn't require a condition. The code in an else block is run if all conditions above that in the <code>if</code> statement evaluate to False.

In [6]:
# a sample state and purchase amount
state = 'CA'
purchase_amount = 20.00    
total_cost = 0

if state == 'CA':
    tax_amount = .075
    total_cost = purchase_amount*(1+tax_amount)

elif state == 'MN':
    tax_amount = .095
    total_cost = purchase_amount*(1+tax_amount)

elif state == 'NY':
    tax_amount = .089
    total_cost = purchase_amount*(1+tax_amount)


print(f"Since you're from {state}, your total cost is ${total_cost:.2f}.")

Since you're from CA, your total cost is $21.50.


In the code above our state was 'CA' so we have a tax ammount of 0.075 and we finished the total_cost with a value of  $21.50

## Quiz 3.1 
### Quiz 3.1 Conditional Statements
Guess My Number

In [ ]:
'''
You decide you want to play a game where you are hiding 
a number from someone.  Store this number in a variable 
called 'answer'.  Another user provides a number called
'guess'.  By comparing guess to answer, you inform the user
if their guess is too high or too low.

# Fill in the conditionals below to inform the user about how
their guess compares to the answer.
'''
answer = #provide answer
guess = #provide guess

if #provide conditional
    result = "Oops!  Your guess was too low."
elif #provide conditional
    result = "Oops!  Your guess was too high."
elif #provide conditional
    result = "Nice!  Your guess matched the answer!"

print(result)

In [ ]:
### Quiz 3.1 Conditional Statements Solution

In [ ]:
answer = 5
guess = 10

if guess<answer:
    result = "Oops!  Your guess was too low."
elif guess>answer:
    result = "Oops!  Your guess was too high."
elif guess == answer:
    result = "Nice!  Your guess matched the answer!"

print(result)